### PR TITLE
AST: Fix Type::join() to do the right thing with class metatypes [3.1]

### PR DESCRIPTION
**Pull request by slavapestov**
Fixes <rdar://problem/30233451>.

### CODE REVIEW

**Comment by slavapestov** [comment]
@DougGregor Is this what you had in mind?

File: lib/AST/TypeJoinMeet.cpp [hunk 1]
```diff
@@ -32,6 +32,18 @@ Type Type::join(Type type1, Type type2) {
...
 
+  // If both are metatypes, compute the join of the instance type and wrap the
+  // result in a metatype.
+  if (auto *metatype1 = type1->getAs<MetatypeType>()) {
+    if (auto *metatype2 = type2->getAs<MetatypeType>()) {
+      auto result = Type::join(metatype1->getInstanceType(),
+                               metatype2->getInstanceType());
+      if (!result)
+        return result;
+      return MetatypeType::get(result);
```
Comment (DougGregor) [inline lib/AST/TypeJoinMeet.cpp]: It's not always true that if `T < U`, then`T.Type < U.Type` in our type system. Let's whitelist the cases where it's true:

* Both instance types `mayHaveSuperclass()`, or
* Both are existential metatypes

and not compute a join in the other cases.

**Review by DougGregor — CHANGES_REQUESTED** [review]
We should narrow this so that it does not produce joins where there isn't actually a subtyping relationship.

File: lib/AST/TypeJoinMeet.cpp (hunk 1 above)
Comment (slavapestov) [inline lib/AST/TypeJoinMeet.cpp]: Ok.

Comment (slavapestov) [inline lib/AST/TypeJoinMeet.cpp]: What's a counter-example? (Non-existential) protocol metatypes?

Comment (DougGregor) [inline lib/AST/TypeJoinMeet.cpp]: `Int.Type` isn't a subtype of `Int?.Type`

**Comment by DougGregor** [comment]
Your revised master PR LGTM!

**Review by DougGregor — APPROVED** [review]

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Pull request merged**